### PR TITLE
Add GriddedInterpolation for non-uniform grids

### PR DESCRIPTION
**Pull request by timholy**
I was looking into doing a benchmark against other packages, and the lack of this feature made the comparisons pretty unfair. So I decided to add it. This was one of those days where one marvels at the efficiency of Julia.

~~I haven't done extensive testing or benchmarking, hence the WIP.~~

### CODE REVIEW

**Comment by timholy** [comment]
OK, I think this is ready to go. Testing could probably be better, but this is a good start.

File: src/b-splines/b-splines.jl [hunk 1]
```diff
@@ -46,18 +46,25 @@ function interpolate{TWeights,TCoefs,IT<:DimSpec{BSpline},GT<:DimSpec{GridType}}
...
+tcoef(A::AbstractArray{Rational{Int}}) = Rational{Int}
+tcoef{T<:Integer}(A::AbstractArray{T}) = typeof(float(zero(T)))
+
 interpolate!{TWeights,IT<:DimSpec{BSpline},GT<:DimSpec{GridType}}(::Type{TWeights}, A, ::Type{IT}, ::Type{GT}) = BSplineInterpolation(TWeights, prefilter!(TWeights, A, IT, GT), IT, GT, Val{0}())
-interpolate!{IT<:DimSpec{BSpline},GT<:DimSpec{GridType}}(A::AbstractArray, ::Type{IT}, ::Type{GT}) = interpolate!(Float64, A, IT, GT)
-interpolate!{IT<:DimSpec{BSpline},GT<:DimSpec{GridType}}(A::AbstractArray{Float32}, ::Type{IT}, ::Type{GT}) = interpolate!(Float32, A, IT, GT)
-interpolate!{IT<:DimSpec{BSpline},GT<:DimSpec{GridType}}(A::AbstractArray{Rational{Int}}, ::Type{IT}, ::Type{GT}) = interpolate!(Rational{Int}, A, IT, GT)
+function interpolate!{IT<:DimSpec{BSpline},GT<:DimSpec{GridType}}(A::AbstractArray, ::Type{IT}, ::Type{GT})
+    interpolate!(tweight(A), A, IT, GT)
+end
```
Comment (tlycken) [inline src/b-splines/b-splines.jl]: This is a nice refactoring! :+1:

File: src/b-splines/indexing.jl [hunk 2]
```diff
@@ -23,14 +23,10 @@ function getindex_impl{T,N,TCoefs,IT<:DimSpec{BSpline},GT<:DimSpec{GridType},Pad
     end
 end
 
-@generated function getindex{T,N}(itp::BSplineInterpolation{T,N}, xs...)
+@generated function getindex{T,N}(itp::BSplineInterpolation{T,N}, xs::Number...)
```
Comment (tlycken) [inline src/b-splines/indexing.jl]: Is this PR based on #54? Didn't we merge this already?

**Comment by tlycken** [comment]
This is really nice!

Re: testing, what is the expected behavior when the knots are actually linspaces? Could we make testing better by comparing a regular b-spline interpolation `bitp = interpolate(A, BSpline{Linear}, OnGrid)` with a gridded `gitp = interpolate(A, Gridded{Linear})` and make sure that their results are within some small epsilon of eachother?

Also, we should start some feature-matrix somewhere, that indicates what features (interpolation degrees, gradient evaluation etc) that are implemented in the different interpolation schemes.

File: src/b-splines/indexing.jl (hunk 2 above)
Comment (timholy) [inline src/b-splines/indexing.jl]: I think I started this branch before that was merged.

**Comment by timholy** [comment]
> Could we make testing better by comparing a regular b-spline interpolation

Good suggestion. This caught what could have been an embarrassing bug.

**Pull request merged**